### ISSUE TITLE
Proposal: Enhance ABI extraction for unverified contracts with EVMole

### ISSUE THREAD

**Issue by cdump** [otterscan/otterscan#2682]
Currently, Otterscan uses WhatsAbi for ABI extraction from unverified contracts, which only provides function selectors without argument information. I'd like to propose integrating (or replacing the current solution with) EVMole, which also extracts function arguments. It can work either on the server side or on the client side inside the browser.

Benchmarks comparing speed and accuracy (including WhatsAbi): https://github.com/cdump/evmole?tab=readme-ov-file#benchmark
JavaScript API documentation: https://github.com/cdump/evmole/tree/master/javascript#api
Online demo: https://evmole.xyz

### Proposed Implementation Flow:
1. Load contract's runtime bytecode
2. Extract function selectors using EVMole (performed offline)
3. Resolve function selectors through https://openchain.xyz/signatures API or any other signature database
4. For any unresolved selectors, use arguments extracted by EVMole

Would love to get your thoughts on this proposal and discuss any implementation details.

**Comment by sealer3** [comment]
Thank you for making a proposal. We are very interested in expanding support for analyzing contracts, and having some way of guessing arguments sounds like a great usability boost. Also, I appreciate the performance benchmarks you made.

I'm also interested in storage slot analysis, e.g. for #2501. From the online demo, it looks like EVMole is able to infer the storage types - how does it do that?

**Comment by cdump** [comment]
> EVMole is able to infer the storage types
Right, this feature was released only 3 days ago, and I haven't published benchmarks yet because the methodology is finally ready, and it's not as simple as for selectors in terms of true positives and false positives. But the accurancy is quite good.


> how does it do that
It works similarly to the selectors and arguments extraction within EVMole. I use the same custom EVM, tracking variables from arguments (also extracted with EVMole) and from some other opcodes (like ADDRESS, BALANCE, etc.). When SSTORE or SLOAD operations occur, I can guess the type. This is a simplified description; the full algorithm can be found here: https://github.com/cdump/evmole/blob/master/src/storage/mod.rs

**Comment by sealer3** [comment]
Very good - I'll definitely be following the progress. Currently, all of the processing in Otterscan is done either in the browser or in the backing Erigon node. We would integrate EVMole on the browser side similar to how we integrate Whatsabi today.

Tracking the storage slots will help us provide a more informative transaction state diff view and might also support a kind of "current contract state" tab. And the enhanced ABI extraction would let us provide a "Read Contract" interface for unverified contracts.
